docs/DesignDocs/openenclave_modules.md

```diff
@@ -0,0 +1,105 @@
+# OpenEnclave Modules
+
+The Open Enclave SDK provides a core set of functionality for build enclave
+applications. Some of these features&2014;such as memory management and ecall/ocall
+marshalling&2014;are required to create an enclave. Other features such as
+libc functions, host filesystem, and network sockets can be optionally included
+inside the enclave by the enclave developer. These optional features are known
+as **Open Enclave Modules** and are static libraries that can be optionally
+linked into the enclave binary.
+
+## Currently supported modules
+
+Currently, Open Enclave only supports modules that are part of the OE SDK source tree.
+Today there are 5 optionally includable modules:
+
+* hostfs
+* libc
+* hostepoll
+* hostsock
+* hostresolver
+
+## Module initialization
+
+OE modules can each have an initialization function which will be called on
+first entry to the enclave. This ensures that a module are fully initialized
+prior to an application using it. There are two ways to demark initialization
+functions:
+
+1. `OE_MOD_INIT Attribute`
+
+The `OE_MOD_INIT` attribute can be applied to a function. This ensures that the
+function is called on enclave entry.
+
+```
+void OE_MOD_INIT init_my_module(void)
```

Comment:
Placing the attribute before the return type would make it consisted with other attributes like OE_NEVER_INLINE.

```c
OE_MODULE_INITIALIZER
void init_my_module(void)
{
    ...
}
```